Nondeterministic "No such file or directory" failure when cleaning project

I'm not sure if this is the right place to report this - I'd rather put this first on some forum, but since SBT does not have any mailing list for users, I'm forced to put it here.

I'm randomly getting this error when trying to build my project:

```
[default-tests-sbt] $ /usr/lib/jvm/java-8-oracle/bin/java -Dsbt.log.noformat=true -Xmx2G -jar /var/lib/jenkins/tools/org.jvnet.hudson.plugins.SbtPluginBuilder_SbtInstallation/SBT_default_launcher/bin/sbt-launch.jar clean test:clean buildUmp test
[info] Loading project definition from /var/lib/jenkins/workspace/default-tests-sbt/project
[info] Set current project to ump (in build file:/var/lib/jenkins/workspace/default-tests-sbt/)
Could not create file /var/lib/jenkins/workspace/default-tests-sbt/launcher/target/streams/$global/ivyConfiguration/$global/streams/outjava.io.IOException: No such file or directory
    at sbt.ErrorHandling$.translate(ErrorHandling.scala:10)
    at sbt.IO$.touch(IO.scala:143)
    at sbt.std.Streams$$anon$3$$anon$2.make(Streams.scala:129)
    at sbt.std.Streams$$anon$3$$anon$2.text(Streams.scala:113)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:124)
    at sbt.std.TaskStreams$class.log(Streams.scala:56)
    at sbt.std.Streams$$anon$3$$anon$2.log$lzycompute(Streams.scala:102)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:102)
    at sbt.Classpaths$$anonfun$mkIvyConfiguration$1.apply(Defaults.scala:1533)
    at sbt.Classpaths$$anonfun$mkIvyConfiguration$1.apply(Defaults.scala:1532)
    at scala.Function10$$anonfun$tupled$1.apply(Function10.scala:35)
    at scala.Function10$$anonfun$tupled$1.apply(Function10.scala:34)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: No such file or directory
    at java.io.UnixFileSystem.createFileExclusively(Native Method)
    at java.io.File.createNewFile(File.java:1012)
    at sbt.IO$$anonfun$1.apply$mcZ$sp(IO.scala:143)
    at sbt.IO$$anonfun$1.apply(IO.scala:143)
    at sbt.IO$$anonfun$1.apply(IO.scala:143)
    at sbt.ErrorHandling$.translate(ErrorHandling.scala:10)
    at sbt.IO$.touch(IO.scala:143)
    at sbt.std.Streams$$anon$3$$anon$2.make(Streams.scala:129)
    at sbt.std.Streams$$anon$3$$anon$2.text(Streams.scala:113)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:124)
    at sbt.std.TaskStreams$class.log(Streams.scala:56)
    at sbt.std.Streams$$anon$3$$anon$2.log$lzycompute(Streams.scala:102)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:102)
    at sbt.Classpaths$$anonfun$mkIvyConfiguration$1.apply(Defaults.scala:1533)
    at sbt.Classpaths$$anonfun$mkIvyConfiguration$1.apply(Defaults.scala:1532)
    at scala.Function10$$anonfun$tupled$1.apply(Function10.scala:35)
    at scala.Function10$$anonfun$tupled$1.apply(Function10.scala:34)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
[error] (launcher/*:ivyConfiguration) Could not create file /var/lib/jenkins/workspace/default-tests-sbt/launcher/target/streams/$global/ivyConfiguration/$global/streams/outjava.io.IOException: No such file or directory
[error] Total time: 76 s, completed 2015-08-14 13:05:08
```

Can anyone direct me how do I even troubleshoot/debug this?

even I am seeing this failure randomly in my sbt builds.

```
Could not create file /home/jenkins/myjH/jobs/dummy-backend-common-build/workspace/source/test-utils/target/streams/$global/projectDescriptors/$global/streams/outjava.io.IOException: No such file or directory
    at sbt.ErrorHandling$.translate(ErrorHandling.scala:10)
    at sbt.IO$.touch(IO.scala:143)
    at sbt.std.Streams$$anon$3$$anon$2.make(Streams.scala:129)
    at sbt.std.Streams$$anon$3$$anon$2.text(Streams.scala:113)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:124)
    at sbt.std.TaskStreams$class.log(Streams.scala:56)
    at sbt.std.Streams$$anon$3$$anon$2.log$lzycompute(Streams.scala:102)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:102)
    at sbt.Classpaths$$anonfun$depMap$1.apply(Defaults.scala:1516)
    at sbt.Classpaths$$anonfun$depMap$1.apply(Defaults.scala:1515)
    at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:35)
    at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:34)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: No such file or directory
    at java.io.UnixFileSystem.createFileExclusively(Native Method)
    at java.io.File.createNewFile(File.java:1012)
    at sbt.IO$$anonfun$1.apply$mcZ$sp(IO.scala:143)
    at sbt.IO$$anonfun$1.apply(IO.scala:143)
    at sbt.IO$$anonfun$1.apply(IO.scala:143)
    at sbt.ErrorHandling$.translate(ErrorHandling.scala:10)
    at sbt.IO$.touch(IO.scala:143)
    at sbt.std.Streams$$anon$3$$anon$2.make(Streams.scala:129)
    at sbt.std.Streams$$anon$3$$anon$2.text(Streams.scala:113)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:124)
    at sbt.std.TaskStreams$class.log(Streams.scala:56)
    at sbt.std.Streams$$anon$3$$anon$2.log$lzycompute(Streams.scala:102)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:102)
    at sbt.Classpaths$$anonfun$depMap$1.apply(Defaults.scala:1516)
    at sbt.Classpaths$$anonfun$depMap$1.apply(Defaults.scala:1515)
    at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:35)
    at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:34)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
[0m[[31merror[0m] [0m(*:[31mprojectDescriptors[0m) Could not create file /home/jenkins/myjH/jobs/dummy-backend-common-build/workspace/source/test-utils/target/streams/$global/projectDescriptors/$global/streams/outjava.io.IOException: No such file or directory[0m
[0m[[31merror[0m] [0mTotal time: 0 s, completed Sep 15, 2015 11:28:08 AM[0m
Build step 'Execute shell' marked build as failure
```

So, I modified my workaround to do this:

```
def tryWorkAroundCrashingSbt(taskKey: TaskKey[_]): Setting[Task[TaskStreams]] = (streams in taskKey) := {
  val fileToCreate = target.value / s"streams/$$global/${taskKey.key.label}/$$global/streams/out"

  var i = 0
  var success = false
  while (!success && i < 100) {
    try {
      IO.touch(fileToCreate)
      success = fileToCreate.exists()
    } catch {
      case e: Exception =>
        streams.value.log.error("Failed to create file " + fileToCreate)
        e.printStackTrace()
    }
    i += 1
  }

  if (!success) {
    throw new Exception("Failed to create streams file " + fileToCreate)
  }
  streams.value.log.info("Created file " + fileToCreate + " after " + i + " trials")

  (streams in taskKey).value
}
```

and it still doesn't work. Here's what's happening now (project names and URLs have been edited):

```
[default-tests] $ /usr/lib/jvm/java-8-oracle/bin/java -Dsbt.log.noformat=true -Xmx2G -jar /var/lib/jenkins/tools/org.jvnet.hudson.plugins.SbtPluginBuilder_SbtInstallation/SBT_default_launcher/bin/sbt-launch.jar clean buildproj test
[info] Loading project definition from <http://rb.company.com/job/default-tests/ws/project>
[info] Set current project to proj (in build <http://rb.company.com/job/default-tests/ws/)>
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-macros/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-macros/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-macros/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv7/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv7/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv7/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-core/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv4/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv4/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv4/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv5/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv5/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv5/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv6/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv6/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv3/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv6/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-shared/target-js/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv8/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-shared/target-jvm/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv2/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-shared/target-js/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv3/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-shared/target-js/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv2/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv8/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv8/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv2/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-doc/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-util/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-shared/target-jvm/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-doc/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-util/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv3/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-shared/target-jvm/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-doc/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[error] Failed to create file <http://rb.company.com/job/default-tests/ws/proj-util/target/streams/$global/ivyConfiguration/$global/streams/out>
Could not create file <http://rb.company.com/job/default-tests/ws/proj-util/target/streams/$global/ivyConfiguration/$global/streams/outjava.io.IOException>: No such file or directory
    at sbt.ErrorHandling$.translate(ErrorHandling.scala:10)
    at sbt.IO$.touch(IO.scala:143)
    at $fbefd52439d4cbd80a5a$$anonfun$tryWorkAroundCrashingSbt$1.apply(build.sbt:67)
    at $fbefd52439d4cbd80a5a$$anonfun$tryWorkAroundCrashingSbt$1.apply(build.sbt:60)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: No such file or directory
    at java.io.UnixFileSystem.createFileExclusively(Native Method)
    at java.io.File.createNewFile(File.java:1012)
    at sbt.IO$$anonfun$1.apply$mcZ$sp(IO.scala:143)
    at sbt.IO$$anonfun$1.apply(IO.scala:143)
    at sbt.IO$$anonfun$1.apply(IO.scala:143)
    ... 21 more
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-util/target/streams/$global/ivyConfiguration/$global/streams/out> after 2 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-core/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-core/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv9/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv9/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-serv9/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-dw/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-launcher/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-launcher/target/streams/$global/ivyConfiguration/$global/streams/out> after 2 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-launcher/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-migration/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-migration/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-dw/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[error] Failed to create file <http://rb.company.com/job/default-tests/ws/proj-mob-core/target/streams/$global/ivyConfiguration/$global/streams/out>
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-migration/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
java.lang.RuntimeException: Could not update last modified time for file <http://rb.company.com/job/default-tests/ws/proj-mob-core/target/streams/$global/ivyConfiguration/$global/streams/out>
    at scala.sys.package$.error(package.scala:27)
    at sbt.IO$.touch(IO.scala:147)
    at $fbefd52439d4cbd80a5a$$anonfun$tryWorkAroundCrashingSbt$1.apply(build.sbt:67)
    at $fbefd52439d4cbd80a5a$$anonfun$tryWorkAroundCrashingSbt$1.apply(build.sbt:60)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-mob-core/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-dw/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-rescl/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-rescl/target/streams/$global/ivySbt/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-mob-core/target/streams/$global/ivyConfiguration/$global/streams/out> after 2 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-rescl/target/streams/$global/ivyConfiguration/$global/streams/out> after 1 trials
[info] Created file <http://rb.company.com/job/default-tests/ws/proj-mob-core/target/streams/$global/projectDescriptors/$global/streams/out> after 1 trials
Could not create file <http://rb.company.com/job/default-tests/ws/proj-rescl/target/streams/$global/projectDescriptors/$global/streams/outjava.io.IOException>: No such file or directory
    at sbt.ErrorHandling$.translate(ErrorHandling.scala:10)
    at sbt.IO$.touch(IO.scala:143)
    at sbt.std.Streams$$anon$3$$anon$2.make(Streams.scala:129)
    at sbt.std.Streams$$anon$3$$anon$2.text(Streams.scala:113)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:124)
    at sbt.std.TaskStreams$class.log(Streams.scala:56)
    at sbt.std.Streams$$anon$3$$anon$2.log$lzycompute(Streams.scala:102)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:102)
    at sbt.Classpaths$$anonfun$depMap$1.apply(Defaults.scala:1516)
    at sbt.Classpaths$$anonfun$depMap$1.apply(Defaults.scala:1515)
    at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:35)
    at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:34)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: No such file or directory
    at java.io.UnixFileSystem.createFileExclusively(Native Method)
    at java.io.File.createNewFile(File.java:1012)
    at sbt.IO$$anonfun$1.apply$mcZ$sp(IO.scala:143)
    at sbt.IO$$anonfun$1.apply(IO.scala:143)
    at sbt.IO$$anonfun$1.apply(IO.scala:143)
    at sbt.ErrorHandling$.translate(ErrorHandling.scala:10)
    at sbt.IO$.touch(IO.scala:143)
    at sbt.std.Streams$$anon$3$$anon$2.make(Streams.scala:129)
    at sbt.std.Streams$$anon$3$$anon$2.text(Streams.scala:113)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:124)
    at sbt.std.TaskStreams$class.log(Streams.scala:56)
    at sbt.std.Streams$$anon$3$$anon$2.log$lzycompute(Streams.scala:102)
    at sbt.std.Streams$$anon$3$$anon$2.log(Streams.scala:102)
    at sbt.Classpaths$$anonfun$depMap$1.apply(Defaults.scala:1516)
    at sbt.Classpaths$$anonfun$depMap$1.apply(Defaults.scala:1515)
    at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:35)
    at scala.Function4$$anonfun$tupled$1.apply(Function4.scala:34)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
[error] (proj-rescl/*:projectDescriptors) Could not create file <http://rb.company.com/job/default-tests/ws/proj-rescl/target/streams/$global/projectDescriptors/$global/streams/outjava.io.IOException>: No such file or directory
[error] Total time: 2 s, completed 2015-09-30 08:51:01
```

It's as if someone deleted the files that I just created...

--- a/main/src/main/scala/sbt/Defaults.scala
+++ b/main/src/main/scala/sbt/Defaults.scala
@@ -374,7 +374,7 @@ object Defaults extends BuildCommon {
   lazy val projectTasks: Seq[Setting[_]] = Seq(
     cleanFiles := cleanFilesTask.value,
     cleanKeepFiles := historyPath.value.toVector,
-    clean := IO.delete(cleanFiles.value),
+    clean := (Def.task { IO.delete(cleanFiles.value) } tag (Tags.Clean)).value,
     consoleProject := consoleProjectTask.value,
     watchTransitiveSources := watchTransitiveSourcesTask.value,
     watch := watchSetting.value
@@ -692,10 +692,15 @@ object Defaults extends BuildCommon {
     Def.task {
       Tests.discover(loadedTestFrameworks.value.values.toList, compile.value, streams.value.log)._1
     }
-  def defaultRestrictions: Initialize[Seq[Tags.Rule]] = parallelExecution { par =>
-    val max = EvaluateTask.SystemProcessors
-    Tags.limitAll(if (par) max else 1) :: Tags.limit(Tags.ForkedTestGroup, 1) :: Nil
-  }
+  def defaultRestrictions: Initialize[Seq[Tags.Rule]] =
+    Def.setting {
+      val par = parallelExecution.value
+      val max = EvaluateTask.SystemProcessors
+      Tags.limitAll(if (par) max else 1) ::
+      Tags.limit(Tags.ForkedTestGroup, 1) ::
+      Tags.exclusiveGroup(Tags.Clean) ::
+      Nil
+    }
 
   lazy val packageBase: Seq[Setting[_]] = Seq(
     artifact := Artifact(moduleName.value)

--- a/main/src/main/scala/sbt/Defaults.scala
+++ b/main/src/main/scala/sbt/Defaults.scala
@@ -372,9 +372,9 @@ object Defaults extends BuildCommon {
   ))
 
   lazy val projectTasks: Seq[Setting[_]] = Seq(
-    cleanFiles := Seq(managedDirectory.value, target.value),
-    cleanKeepFiles := historyPath.value.toList,
-    clean := doClean(cleanFiles.value, cleanKeepFiles.value),
+    cleanFiles := cleanFilesTask.value,
+    cleanKeepFiles := historyPath.value.toVector,
+    clean := IO.delete(cleanFiles.value),
     consoleProject := consoleProjectTask.value,
     watchTransitiveSources := watchTransitiveSourcesTask.value,
     watch := watchSetting.value
@@ -816,14 +816,21 @@ object Defaults extends BuildCommon {
     pickMainClass(classes)
   }
 
-  def doClean(clean: Seq[File], preserve: Seq[File]): Unit =
-    IO.withTemporaryDirectory { temp =>
-      val (dirs, files) = preserve.filter(_.exists).flatMap(_.allPaths.get).partition(_.isDirectory)
-      val mappings = files.zipWithIndex map { case (f, i) => (f, new File(temp, i.toHexString)) }
-      IO.move(mappings)
-      IO.delete(clean)
-      IO.createDirectories(dirs) // recreate empty directories
-      IO.move(mappings.map(_.swap))
+  /** Implements `cleanFiles` task. */
+  def cleanFilesTask: Initialize[Task[Vector[File]]] =
+    Def.task {
+      val filesAndDirs = Vector(managedDirectory.value, target.value)
+      val preserve = cleanKeepFiles.value
+      val (dirs, fs) = filesAndDirs.filter(_.exists).partition(_.isDirectory)
+      val preserveSet = preserve.filter(_.exists).toSet
+      // performance reasons, only the direct items under `filesAndDirs` are allowed to be preserved.
+      val dirItems = dirs flatMap { _.glob("*").get }
+      (preserveSet diff dirItems.toSet) match {
+        case xs if xs.isEmpty => ()
+        case xs               => sys.error(s"cleanKeepFiles contains directory/file that are not directly under cleanFiles: $xs")
+      }
+      val toClean = (dirItems filterNot { preserveSet(_) }) ++ fs
+      toClean
     }
 
   def bgRunMainTask(products: Initialize[Task[Classpath]], classpath: Initialize[Task[Classpath]],

--- a/main/src/main/scala/sbt/Keys.scala
+++ b/main/src/main/scala/sbt/Keys.scala
@@ -162,8 +162,8 @@ object Keys {
 
   // Output paths
   val classDirectory = SettingKey[File]("class-directory", "Directory for compiled classes and copied resources.", AMinusSetting)
-  val cleanFiles = SettingKey[Seq[File]]("clean-files", "The files to recursively delete during a clean.", BSetting)
-  val cleanKeepFiles = SettingKey[Seq[File]]("clean-keep-files", "Files to keep during a clean.", CSetting)
+  val cleanFiles = TaskKey[Seq[File]]("clean-files", "The files to recursively delete during a clean.", BSetting)
+  val cleanKeepFiles = SettingKey[Seq[File]]("clean-keep-files", "Files or directories to keep during a clean. Must be direct children of target.", CSetting)
   val crossPaths = SettingKey[Boolean]("cross-paths", "If true, enables cross paths, which distinguish input and output directories for cross-building.", ASetting)
   val taskTemporaryDirectory = SettingKey[File]("task-temporary-directory", "Directory used for temporary files for tasks that is deleted after each task execution.", DSetting)
 

--- a/main/src/main/scala/sbt/Tags.scala
+++ b/main/src/main/scala/sbt/Tags.scala
@@ -13,6 +13,7 @@ object Tags {
   val Test = Tag("test")
   val Update = Tag("update")
   val Publish = Tag("publish")
+  val Clean = Tag("clean")
 
   val CPU = Tag("cpu")
   val Network = Tag("network")
@@ -24,7 +25,7 @@ object Tags {
    * Describes a restriction on concurrently executing tasks.
    * A Rule is constructed using one of the Tags.limit* methods.
    */
-  sealed trait Rule { // TODO: make this an abstract class for 0.14
+  abstract class Rule {
     def apply(m: TagMap): Boolean
     def ||(r: Rule): Rule = new Or(this, r)
     def &&(r: Rule): Rule = new And(this, r)

